gateway: add an authenticating reverse proxy

anisometropie opened the pull request:
Before this change, we relied on an authenticating reverse proxy called [gateway](https://github.com/pguenezan/gateway). It's been in use since the inception of the project, as part of a package of services we got from SNCF Réseau.

The front-end was responsible for authenticating with keycloak, and providing a Bearer token for the reverse proxy to check.

This change introduces an authenticating reverse proxy, which replaces the previous gateway.
The front-end now entirely delegates authentication to the backend.

On startup, `gateway/auth/login` is called, which either returns an username
if the client is authenticated, or a redirection URL if the user needs to log in.

There are three authentication backends supported by the new gateway:
 - A default mock backend, which return all users to be authentificated with the same
   username
 - An OpenID Connect backend, which is used to authenticate interactive users.
   The application does not support session refresh, and does not forward logout events to the
   identity provider.
 - A static Bearer token backend, which is not meant to be used by interactive users

Fixes #1524
Fixes #2349
Fixes #5339

flomonster commented on editoast/src/main.rs:
```diff
@@ -118,6 +124,56 @@ fn init_sentry(args: &RunserverArgs) -> Option<ClientInitGuard> {
+    credentials: BearerAuth,
+) -> Result<ServiceRequest, (ActixError, ServiceRequest)> {
+    // let secret = env::var("OSRD_AUTH_ZERO_SECRET").expect("auth0 secret must be set");
+    let domain = env::var("OSRD_AUTH_ZERO_DOMAIN").expect("auth0 domain must be set");
```
- We should use clap to retrieve env var or app options
- Rename to `OIDC_JWK_URIS`
  - This can contain multiple URI (not domains)

flomonster commented on editoast/src/main.rs:
```diff
@@ -118,6 +124,56 @@ fn init_sentry(args: &RunserverArgs) -> Option<ClientInitGuard> {
+    keys: Vec<Jwk>,
+}
+
+async fn get_jwks(domain: &str) -> Result<Vec<Jwk>, reqwest::Error> {
+    let jwks_uri = format!("{}.well-known/jwks.json", domain);
+    let jwks: JwksResponse = reqwest::get(&jwks_uri).await?.json().await?;
+    Ok(jwks.keys)
+}
```
We can improve this function by:
- Having a list of `jwks_uri` as the argument
- Don't use a domain but full URI (What you did is specific to auth0)
- This function should return a `jsonwebtoken::jwk::JwkSet`.
- To do so, you can deserialize multiple JwkSet and then merge them into a single one.

flomonster commented on editoast/src/main.rs:
```diff
@@ -118,6 +124,56 @@ fn init_sentry(args: &RunserverArgs) -> Option<ClientInitGuard> {
+
+    let jwks = get_jwks(&domain).await.expect("Failed to fetch JWKS");
+    // let decoding_key = DecodingKey::from_secret(secret.as_ref());
+    let decoding_key = DecodingKey::from_rsa_components(&jwks[0].n, &jwks[0].e).unwrap();
+
+    match decode::<Claims>(&credentials.token(), &decoding_key, &validation) {
```
We need to know which key to use before decoding the token.

- Use [jsonwebtoken::decode_header](https://docs.rs/jsonwebtoken/latest/jsonwebtoken/fn.decode_header.html#) to retrieve the `kid` of the token. This will retrieve the public key associated with the private key used to sign the token.
- Use the [find method](https://docs.rs/jsonwebtoken/latest/jsonwebtoken/jwk/struct.JwkSet.html#method.find) to retrieve the Jwk.
- Use the `from_jwk` method. Check documentation: https://docs.rs/jsonwebtoken/latest/jsonwebtoken/struct.DecodingKey.html#method.from_jwk. (This will automatically check the algorithm used by the public key.)
- Then call the decode function.

flomonster commented on editoast/src/main.rs:
```diff
@@ -150,6 +206,8 @@ async fn runserver(
 
     let server = HttpServer::new(move || {
+        let auth = HttpAuthentication::bearer(validator);
+        let use_auth = env::var("OSRD_BACKEND_ENVIRONMENT").map_or(false, |v| v == "sncf_reseau");
```
Instead of having another env var. We can simply check if the `OIDC_JWK_URIS` is given.

flomonster commented on editoast/src/main.rs:
```diff
@@ -118,6 +124,56 @@ fn init_sentry(args: &RunserverArgs) -> Option<ClientInitGuard> {
+    let mut validation = Validation::new(Algorithm::RS256);
+    validation.set_issuer(&[&domain]);
+
+    let jwks = get_jwks(&domain).await.expect("Failed to fetch JWKS");
```
We should cache the `JwkSet` at startup not in the validator.
To retrieve the `JkwSet` we can use an app_data. Checkout this example https://docs.rs/actix-web-httpauth/0.8.0/actix_web_httpauth/middleware/struct.HttpAuthentication.html#examples-1

flomonster reviewed (commented):
Here is an intermediate review.

flomonster commented on editoast/src/client/mod.rs:
```diff
@@ -77,6 +77,11 @@ pub struct RunserverArgs {
     #[derivative(Default(value = r#""".into()"#))]
     #[clap(long, env = "ROOT_PATH", default_value_t = String::new())]
     pub root_path: String,
+    /// A comma separated list of URIs from which are fetched the public keys
+    /// If present, authentication is enable and editoast will require a valid Bearer token for every call.
+    /// If absent, no authentication is required.
+    #[clap(long, env = "OIDC_JWK_URIS")]
+    pub oidc_jwk_uris: Option<String>,
```
Can use value delimiter: https://docs.rs/clap/latest/clap/struct.Arg.html#method.value_delimiter

```suggestion
    /// A comma separated list of URIs from which are fetched the public keys
    /// If present, authentication is enable and editoast will require a valid Bearer token for every call.
    /// If absent, no authentication is required.
    #[clap(long, env = "OIDC_JWK_URIS")]
    #[arg(value_delimiter(','))] 
    pub oidc_jwk_uris: Vec<String>,
```

flomonster commented on editoast/src/main.rs:
```diff
@@ -134,6 +141,70 @@ fn log_received_request(req: &ServiceRequest) {
+) -> Result<ServiceRequest, (ActixError, ServiceRequest)> {
+    let jwt_header = decode_header(&credentials.token()).unwrap();
+    let token_kid = jwt_header.kid.ok_or("No kid in token").unwrap();
+    let jwks = get_jwks(oidc_jwk_uris, &req)
```
Retrieve the `jkws` should be done at startup.

flomonster commented on editoast/src/main.rs:
```diff
@@ -134,6 +141,70 @@ fn log_received_request(req: &ServiceRequest) {
+}
+
+async fn validator(
+    oidc_jwk_uris: Option<String>,
```
This field won't be needed.

flomonster replied:
```suggestion
    oidc_jwk_uris: Vec<String>,
```

flomonster commented on editoast/src/main.rs:
```diff
@@ -211,7 +292,7 @@ async fn runserver(
     info!("Running server...");
     server
-        .bind((args.address.clone(), args.port))?
+        .bind((args.address.clone(), args.port.clone()))?
```
```suggestion
        .bind((args.address.clone(), args.port))?
```

flomonster commented on editoast/src/main.rs:
```diff
@@ -134,6 +141,70 @@ fn log_received_request(req: &ServiceRequest) {
+    req: ServiceRequest,
+    credentials: BearerAuth,
+) -> Result<ServiceRequest, (ActixError, ServiceRequest)> {
+    let jwt_header = decode_header(&credentials.token()).unwrap();
+    let token_kid = jwt_header.kid.ok_or("No kid in token").unwrap();
```
We should transform the error in `(ActixError, ServiceRequest)` instead of `unwrap`.

flomonster commented on editoast/src/main.rs:
```diff
@@ -134,6 +141,70 @@ fn log_received_request(req: &ServiceRequest) {
+        Some(jwk) => jwk,
+        None => {
+            return Err((
+                ErrorForbidden("no jwk found for the given kid".to_string()),
```
```suggestion
                ErrorForbidden(format!("Kid '{}' not found in key set", token_kid)),
```

flomonster commented on editoast/src/main.rs:
```diff
@@ -134,6 +141,70 @@ fn log_received_request(req: &ServiceRequest) {
+        }
+    };
+
+    let validation = Validation::new(Algorithm::RS256);
+    let decoding_key = DecodingKey::from_jwk(jwk).unwrap();
```
```suggestion
    let validation = Validation::new(jwt_header.alg);
    let decoding_key = DecodingKey::from_jwk(jwk).unwrap();
```

flomonster commented on editoast/src/main.rs:
```diff
@@ -134,6 +141,70 @@ fn log_received_request(req: &ServiceRequest) {
+            Err((actix_error, req))
+        }
+    }
+    // }
```
```suggestion
```

flomonster commented on editoast/src/main.rs:
```diff
@@ -185,6 +264,7 @@ async fn runserver(
                 is_sentry_initialized,
                 sentry_actix::Sentry::new(),
             ))
+            .wrap(Condition::new(use_auth, auth.clone()))
```
```suggestion
```

flomonster commented on editoast/src/main.rs:
```diff
@@ -161,10 +232,15 @@ async fn runserver(
     // Setup shared states
     let infra_caches = Data::new(CHashMap::<i64, InfraCache>::default());
 
+    let oidc_public_keys_cache = Data::new(Mutex::new(JwkSet { keys: Vec::new() }));
```
- We can retrieve the key set at startup
- No mutex needed (read only)

flomonster commented on editoast/src/main.rs:
```diff
@@ -134,6 +141,70 @@ fn log_received_request(req: &ServiceRequest) {
+            .split(',')
+            .map(|s| s.trim().to_string()) // trim to remove any unwanted spaces
+            .collect();
+        println!("{:?}", oidc_jwk_uris);
```
```suggestion
```

flomonster commented on editoast/src/main.rs:
```diff
@@ -134,6 +141,70 @@ fn log_received_request(req: &ServiceRequest) {
+async fn get_jwks(
+    oidc_jwk_uris: Option<String>,
+    req: &ServiceRequest,
+) -> Result<JwkSet, reqwest::Error> {
```
```suggestion
async fn update_jwk_set(
    oidc_jwk_uris: Vec<String>,
    req: &ServiceRequest,
) -> Result<JwkSet, reqwest::Error> {
```

flomonster commented on editoast/src/main.rs:
```diff
@@ -134,6 +141,70 @@ fn log_received_request(req: &ServiceRequest) {
+) -> Result<JwkSet, reqwest::Error> {
+    let cache_mutex = req.app_data::<Data<Mutex<JwkSet>>>().unwrap();
+    let mut cache = cache_mutex.lock().unwrap();
+    if cache.keys.is_empty() {
```
This function should update the cached `jwk_set`.
```suggestion
```

flomonster commented on editoast/src/main.rs:
```diff
@@ -134,6 +141,70 @@ fn log_received_request(req: &ServiceRequest) {
+) -> Result<ServiceRequest, (ActixError, ServiceRequest)> {
+    let jwt_header = decode_header(&credentials.token()).unwrap();
+    let token_kid = jwt_header.kid.ok_or("No kid in token").unwrap();
+    let jwks = get_jwks(oidc_jwk_uris, &req)
+        .await
```
Here we should retrieve the cached `jwk_set` and not update it.

flomonster commented on editoast/src/main.rs:
```diff
@@ -134,6 +141,70 @@ fn log_received_request(req: &ServiceRequest) {
+        .await
+        .unwrap_or(JwkSet { keys: vec![] });
+    let jwk_to_use = jwks.find(&token_kid);
+    let jwk = match jwk_to_use {
```
If the `Jwk` is not found in the set we should update the set and try again. If we don't find it again then we can return an error.

bloussou commented on docker-compose-host.yml:
```diff
@@ -60,7 +60,7 @@ services:
         OSRD_GIT_DESCRIBE: ${OSRD_GIT_DESCRIBE}
     environment:
-    - OSRD_LOCAL_BACKEND=true
+    - OSRD_BACKEND_ENVIRONMENT=local
```
This is not about local backend, it is with or without authentication

multun committed:
gateway: add an authenticating reverse proxy

Before this change, we relied on an authenticating reverse proxy called gateway. It's been in use since the inception of the project, as part of a package of services we got from SNCF Réseau.

The front-end was responsible for authenticating with keycloak, and providing a Bearer token for the reverse proxy to check.

This change introduces an authenticating reverse proxy, which replaces the previous gateway.
The front-end now entirely delegates authentication to the backend.

On startup, gateway/auth/login is called, which either returns an username
if the client is authenticated, or a redirection URL if the user needs to log in.

There are three authentication backends supported by the new gateway:

- A default mock backend, which return all users to be authentificated with the same
  username
- An OpenID Connect backend, which is used to authenticate interactive users.
  The application does not support session refresh, and does not forward logout events to the
  identity provider.
- A static Bearer token backend, which is not meant to be used by interactive users

This change also introduces unstable builds, which are pushed to ghcr.io.

Co-Authored-By: Valentin Chanas <anisometropie@gmail.com>
Co-Authored-By: =?UTF-8?q?=C3=89lys=C3=A6?= <101974839+ElysaSrc@users.noreply.github.com>

leovalais commented on editoast/src/auth/mod.rs:
Endpoints are (for now at least) all located in `views`. I'm not against breaking that and having each endpoint next to its feature definition since we're talking about splitting editoast into several crates ; but I'd like to have a second opinion on that.

leovalais commented on editoast/src/auth/mod.rs:
```diff
@@ -0,0 +1,345 @@
+    request: HttpRequest,
+    session: Session,
+    auth_state: Data<Mutex<AuthState>>,
+) -> Result<web::Json<AuthResponse>, InternalError> {
```
For consistency's sake use `editoast::error::Result<T>`.

leovalais reviewed (commented):
Thanks! I've read the code, lgtm, left a few comments. I'll approve once I've tested it.

clarani commented on front/README.md:
```diff
@@ -21,6 +21,12 @@ well.
+### `yarn start-https`
+
+Runs the app in local development environement with https.
+to generate a certificate that you can use locally, run `yarn make-certificate".
```
Add capital letter to "To" and change " => `
```suggestion
To generate a certificate that you can use locally, run `yarn make-certificate`.
```

clarani commented on front/README.md:
```diff
@@ -21,6 +21,12 @@ well.
+
+Runs the app in local development environement with https.
+to generate a certificate that you can use locally, run `yarn make-certificate".
+You’ll need libnss3-tools and mkcert installed
```
```suggestion
You’ll need libnss3-tools and mkcert installed.
```

leovalais reviewed (approved):
Tested authenticating using SNCF test OIDC provider and without authentication.

clarani commented on front/src/common/api/documentApi.ts:
```diff
@@ -1,12 +1,10 @@
 
 export const getDocument = async (documentKey: number): Promise<Blob> => {
     const config = {
-        headers: {
-          Authorization: `Bearer ${localStorage.getItem('access_token')}`,
-        },
```
Did you forget the `credentials: "include"` here ?

clarani commented on front/src/reducers/user/index.ts:
```diff
@@ -1,118 +1,44 @@
-  username: string;
-  accessToken?: string;
+  loginError?: ApiError;
+  username?: string;
```
Username cannot be undefined since by default it is an empty string
```suggestion
  username: string;
```

clarani commented on front/src/reducers/user/userReducer.spec.ts:
```diff
@@ -1,104 +1,69 @@
-  test('should handle logoutUser', () => {
+  it('should handle logoutUser', () => {
     const store = createStore({
+      ...userInitialState,
```
Can you add a username not empty here ?
```suggestion
      ...userInitialState,
      username: 'test logout'
```

clarani commented on front/src/vite-env.d.ts:
```diff
@@ -3,7 +3,7 @@
-  OSRD_KEYCLOAK_REALM: string;
   OSRD_SENTRY_DSN: string;
   OSRD_SENTRY_ENVIRONMENT: string;
+  OSRD_GIT_DESCRIBE: string;
```
What is it used for ?

clarani reviewed (commented):
Left some comments

clarani commented on front/vite.config.ts:
```diff
@@ -31,6 +33,12 @@ export default defineConfig(({ mode }) => {
       outDir: 'build',
     },
     server: {
+      ...(process.env.USE_HTTPS === 'true' && {
```
Add USE_HTTPS in vite-env too

clarani commented on front/src/utils/helpers.ts:
```diff
@@ -72,7 +72,7 @@ export function transformMapRequest(
     return {
       url,
-      headers: { Authorization: `Bearer ${localStorage.getItem('access_token')}` },
+      headers: { credentials: 'include' },
```
:warning: it does not work with React MapGL :/ 

![image](https://github.com/osrd-project/osrd/assets/45000526/b07071dc-90c1-42c0-a201-9874c03d110c)

clarani commented on front/src/common/requests.ts:
:warning: You should add the new header each time the functions of this file are used : 
- editor/data/api: getEntities, getRoutesFromWaypoint, getAttachedItems, getRouteTrackRanges, getCompatibleRoutes

Khoyo commented on gateway/import_meta_env/Cargo.toml:
```diff
@@ -0,0 +1,16 @@
+actix-service = "2"
+serde = "1.0"
+serde_json = "1.0"
+futures-util = "0.3"
```
Missing newline at EOF

Khoyo commented on docker/entrypoint.sh:
```diff
@@ -0,0 +1,11 @@
+#!/bin/sh
```
maybe call that file `gateway_entrypoint.sh`

Khoyo commented on docker/gateway-embedded-front.dockerfile:
```diff
@@ -0,0 +1,57 @@
+# Build dependencies
+COPY package.json yarn.lock /app/
+RUN yarn
+COPY . /app
```
missing `--from`

Khoyo commented on docker/gateway-embedded-front.dockerfile:
```diff
@@ -0,0 +1,57 @@
+# Prepare the recipe
+FROM chef as planner
+COPY . .
+RUN cargo chef prepare --recipe-path recipe.json
```
missing `--from`

Khoyo commented on gateway/actix_auth/src/auth_context.rs:
```diff
@@ -0,0 +1,156 @@
+#[derive(Clone)]
+pub struct AuthContext {
+    session_prefix: String,
+    identity_providers: Vec<(String, Box<dyn IdentityProvider>)>,
```
Is this supposed to be a `Vec`?

multun replied:
yes

Khoyo commented on gateway/actix_auth/src/auth_context.rs:
```diff
@@ -0,0 +1,156 @@
+                return SessionStatus::LoggedOut;
+            }
+            Some(session_provider) => session_provider,
+        };
```
Those can be let-else

Khoyo commented on gateway/actix_auth/src/auth_context.rs:
```diff
@@ -0,0 +1,156 @@
+        let session = match session {
+            None => return IdentityStatus::Unknown,
+            Some(session) => session,
+        };
```
let-else, or `ok_or(IdentityStatus::Unknown)?`

Khoyo commented on gateway/actix_auth/src/providers/oidc.rs:
```diff
@@ -0,0 +1,268 @@
+            .add_scope(Scope::new("profile".to_owned()));
+
+        let (auth_url, csrf_token, nonce) = client_builder.url();
+        // if we decide ever decide to store the nonce unencrypted, we should also change how it's checked:
```
decide ever decide

Khoyo commented on gateway/actix_proxy_service/src/lib.rs:
```diff
@@ -0,0 +1,315 @@
+    query: &str,
+    new_protocol: Option<&str>,
+) -> Uri {
+    let scheme = new_protocol.unwrap_or(target.upstream.scheme_str().unwrap());
+    let authority = target.upstream.authority().unwrap().clone();
```
Please use `expect()` instead of `unwrap()`. Also, are we sure to always have an authority ?

ElysaSrc replied:
We do always have an authority, it's checked when the config file is parsed.

Khoyo commented on gateway/actix_proxy_service/src/lib.rs:
```diff
@@ -0,0 +1,315 @@
+        .authority(authority)
+        .path_and_query(final_path)
+        .build()
+        .unwrap()
```
Please use expect() instead of unwrap()

flomonster commented on front/.env.example:
```diff
@@ -1,7 +1,5 @@
-OSRD_LOCAL_BACKEND=false
 OSRD_API_URL=http://example.com/api
```
`OSRD_API_URL` can be drop too.

flomonster commented on docker/gateway.prod.toml:
```diff
@@ -0,0 +1,18 @@
+filename = "index.html"
+also_serve_root = true
+env_vars = [
+  "OSRD_API_URL", 
```
```suggestion
```
This was the legacy url to the `api` service.

flomonster commented on .github/workflows/docker-build-push.yml:
```diff
@@ -22,6 +22,21 @@ jobs:
+      DOCKERHUB_USERNAME: ${{ secrets.DOCKERHUB_USERNAME }}
+      DOCKERHUB_TOKEN: ${{ secrets.DOCKERHUB_TOKEN }}
+
   build_push_front:
```
Do we want to keep building this image?

multun replied:
probably not, it's gone now

flomonster reviewed (commented):
Awesome job! :tada:

Small to-do list:
- Add CODEOWNERS (`osrd-project/devops`)
- Use workspace dependencies
- Add a `README.md`
- Add a workflow to run clippy and fmt.

Castavo commented on gateway/src/config.rs:
```diff
@@ -0,0 +1,130 @@
+    /// The request path must start with this prefix for the target to apply.
+    /// If omitted, the target becomes the default. There can only be a single default target.
+    pub prefix: Option<String>,
+    /// The base URL to requests are proxied to (must include the scheme)
```
```suggestion
    /// The base URL the requests are proxied to (must include the scheme)
```

flomonster commented on gateway/src/config.rs:
```diff
@@ -0,0 +1,130 @@
+    },
+}
+
+#[derive(Deserialize, Serialize, Clone)]
```
We could add the `deny_unknown_fields`.

```suggestion
#[derive(Deserialize, Serialize, Clone)]
#[serde(deny_unknown_fields)]
```

bloussou reviewed (changes requested):
Refresh is impossible in the front when you are not on the base url 
![image](https://github.com/osrd-project/osrd/assets/32448554/fa8d12db-bd57-4e29-a3e9-612644dd9b9c)

bloussou commented:
We should add a `/health` endpoint

flomonster commented on gateway/src/config.rs:
```diff
@@ -0,0 +1,130 @@
+
+#[derive(Deserialize, Serialize, Clone)]
+pub struct AuthConfig {
+    pub session_namespace: String,
```
What is this setting used for?

flomonster commented on gateway/src/config.rs:
```diff
@@ -0,0 +1,114 @@
+        require_login: bool,
+    },
+
+    /// Basic authentication provider
+    /// This provider will check the username and password against a list of
+    /// username/password pairs
+    Bearer {
+        provider_id: String,
+        tokens: HashMap<String, String>,
+    },
```
I find the documentation confusing.
Why are we talking about passwords, we should talk about tokens instead. In addition, we should specify that the map is `token` -> `username` and not the other way around.

flomonster replied:
Actually, we could change the implementation to `username` -> `token` with an error if multiple users have the same token :shrug:

ElysaSrc replied:
Changed documentation accordingly

Castavo commented on gateway/src/main.rs:
```diff
@@ -0,0 +1,98 @@
+        .map(|s| s.parse().expect("Invalid trusted proxy IP CIDR"))
+        .collect::<Vec<_>>();
+
+    let auth_context = parse_auth_config(config.clone().auth).await;
```
```suggestion
    let auth_context = parse_auth_config(config.auth.clone()).await;
```

Castavo commented on gateway/README.md:
It would be nice to have an example or explanations on how the files option is supposed to be used

Castavo commented on gateway/actix_auth/src/gate_middleware.rs:
It feels like this file isn't used :eyes:

Castavo replied:
I took too long to post this review :grimacing: 
The file is still there and still feels unused tho

flomonster commented on gateway/actix_proxy/src/lib.rs:
```diff
@@ -0,0 +1,396 @@
+    }
+}
+
+/// Convert `SendRequestError` to a server `Response`
+///
```
```suggestion
/// Convert `SendRequestError` to a server `Response`
```

flomonster reviewed (approved):
Very impressive PR. Thanks!
Note: Reviewed and tested with different config files on NixOS.

clarani commented on front/docker/dev-entrypoint.sh:
```diff
@@ -0,0 +1,22 @@
+new_uid=$(stat -c %u .)
+new_gid=$(stat -c %g .)
+
+# don't create an user, as it the UID may already
```
```suggestion
# don't create an user, as the UID may already
```

clarani commented on front/src/utils/hooks/OsrdAuth.ts:
```diff
@@ -0,0 +1,38 @@
+  return {
+    username,
+    isUserLogged,
+    isLocalBackend: false,
```
Do you need this ? I think you don't need it anymore, as it is always false. You can also remove its usage in App.jsx

clarani reviewed (commented):
Reviewed all the code in /front, I still need to test but it will wait till Monday :) Left 2 small comments !

alexandredamiron reviewed (changes requested):
gateway timeout results in 500. 5s should ne be enough for multiple configurations

clarani reviewed (approved):
LGTM, tested in local (with docker and with the front+gateway+editoast in local) :white_check_mark:

Castavo reviewed (commented):
I tested it while running all services locally, the proxy works fine ! 

When running with the mock auth provider, I can't logout from the front however, I get `[2023-11-14T16:48:50Z WARN  actix_proxy] proxy: error forwarding request: authentication required`

leovalais reviewed (approved):
Tested on mac arm using Safari and Chrome, both locally and with docker.
